highfidelity/hifi
# Commerce: staticCertificateVersion; Seal serverScripts JSON property; getStaticCertificateJSON()

This PR accomplishes three things:
1. It adds support for static certificate JSON versioning, which is added on the backend [here](https://github.com/highfidelity/data-web/pull/1258)
2. It makes use of the `serverScripts` JSON property when performing static cert verification for entities with a static cert JSON version of > 0
3. It adds a `getStaticCertificateJSON()` endpoint to the `Entities` scripting interface. This returns a string (in JSON format) that is equal to the static certificate JSON.

**Test Plan:**

1. Install this PR (Interface and Sandbox).
2. Create an Environment Variable on your system called `HIFI_METAVERSE_URL` and set its value to `https://staging.highfidelity.com`
    - For instructions about doing this on Windows 10, [click here](https://superuser.com/questions/284342/what-are-path-and-other-environment-variables-and-how-can-i-set-or-use-them)
3. Run Sandbox and Interface. Open Interface, log in to your Staging account (creating one if necessary), and connect to your Sandbox.
    - To avoid confusion, I recommend your Staging username be different from your Stable username - for example, I use `zfox` on Stable and `zfoxStaging` on Staging
4. If you don't have a Wallet set up already, set up a Wallet. If you haven't already requested seeding on the Staging metaverse, request seeding from me via Slack.
5. Purchase and rez the `Weaponized Garden Gnome` from the Marketplace. Verify that you are able to purchase and rez the gnome without issues.
6. Get the Entity ID of the Gnome using Create mode, then open your JavaScript console (CTRL+SHIFT+J) and type: `Entities.getStaticCertificateJSON("<Gnome Entity ID>");`. Verify that the returned string appears reasonable and **does NOT** contain a `serverScripts` key/value pair.
7. Export the rezzed `Weaponized Garden Gnome` from your domain:
    1. Open Create
    2. Click the gnome
    3. Go to `Edit` -> `Export Entities`
    4. Export the JSON to your desktop (or some other memorable place)
8. Using your favorite text editor, open the JSON file you exported in (7).
9. Add a new key/value pair to the JSON file: `"serverScripts": "http://example.com",`, then save the file.
10. Move your view from within Interface slightly, then import the JSON you just edited in (9).
11. Verify that the initial Weaponized Garden Gnome that you rezzed in (5) has disappeared.
12. Verify that the new Weaponized Garden Gnome that you rezzed in (10) is present in your Sandbox.
13. Get the Entity ID of the Gnome using Create mode, then open your JavaScript console (CTRL+SHIFT+J) and type: `Entities.getStaticCertificateJSON("<Gnome Entity ID>");`. Verify that the returned string appears reasonable and **does NOT** contain a `serverScripts` key/value pair.
14. Purchase and rez the `2017-12-14 Beach Ball with serverScripts` from the Marketplace. Verify that you are able to purchase and rez the ball without issues.
15. Get the Entity ID of the Beach Ball using Create mode, then open your JavaScript console (CTRL+SHIFT+J) and type: `Entities.getStaticCertificateJSON("<Beach Ball Entity ID>");`. Verify that the returned string appears reasonable and **does** contain a `serverScripts` key/value pair.
16. Export the rezzed `Beach Ball with serverScript` from your domain **TWICE** (so that you have two identical copies of the JSON):
    1. Open Create
    2. Click the beach ball
    3. Go to `Edit` -> `Export Entities`
    4. Export the JSON to your desktop (or some other memorable place)
    5. Repeat 3-4 one more time
17. Using your favorite text editor, open the 1st copy of the JSON file you exported in (16).
18. Find the value associated with the key `serverScripts`. Modify it to be `http://fake.com`, then save the file.
19. Move your view from within Interface slightly, then import the JSON you just edited in (18).
20. Verify that the initial Beach Ball that you rezzed in (14) has disappeared.
21. Verify that the new Beach Ball that you attempted to rez in (19) is **NOT** present in your Sandbox.
22. Using your favorite text editor, open the 2nd copy of the JSON file you exported in (16).
23. Find the value associated with the key `staticCertificateVersion`. Remove that key/value pair, then save the file.
24. Move your view from within Interface slightly, then import the JSON you just edited in (23).
25. Verify that the new Beach Ball that you attempted to rez in (24) is **NOT** present in your Sandbox.
